Index: src/genplusgx/loadrom.c
===================================================================
--- src/genplusgx/loadrom.c	(revision 436)
+++ src/genplusgx/loadrom.c	(revision 437)
@@ -238,11 +238,10 @@
 }
 
 /* SMD (interleaved) rom support */
-static uint8 block[0x4000];
-
 static void deinterleave_block (uint8 * src)
 {
   int i;
+  uint8 block[0x4000];
   memcpy (block, src, 0x4000);
   for (i = 0; i < 0x2000; i += 1)
   {
