Index: src/genplusgx/loadrom.h
===================================================================
--- src/genplusgx/loadrom.h	(revision 442)
+++ src/genplusgx/loadrom.h	(revision 443)
@@ -64,6 +64,7 @@
 extern PERIPHERALINFO peripheralinfo[14];
 extern uint16 realchecksum;
 extern uint16 peripherals;
+extern char rom_filename[256];
 
 /* Function prototypes */
 extern int load_rom(char *filename);
Index: src/genplusgx/loadrom.c
===================================================================
--- src/genplusgx/loadrom.c	(revision 442)
+++ src/genplusgx/loadrom.c	(revision 443)
@@ -60,7 +60,9 @@
 uint16 peripherals;
 uint16 realchecksum;
 ROMINFO rominfo;
+char rom_filename[256];
 
+
 /***************************************************************************
   * Genesis ROM Manufacturers
   *
