Index: src/genplusgx/state.c
===================================================================
--- src/genplusgx/state.c	(revision 529)
+++ src/genplusgx/state.c	(revision 530)
@@ -22,7 +22,7 @@
 
 #include "shared.h"
 
-#define STATE_VERSION "GENPLUS-GX 1.4.x"
+#define STATE_VERSION "GENPLUS-GX 1.4.0"
 
 #define load_param(param, size) \
   memcpy(param, &state[bufferptr], size); \
Index: src/genplusgx/gx/osd.h
===================================================================
--- src/genplusgx/gx/osd.h	(revision 529)
+++ src/genplusgx/gx/osd.h	(revision 530)
@@ -33,9 +33,9 @@
 #define SK_UPMEM      "/genplus/sk2chip.bin"
 
 #ifdef HW_RVL
-#define VERSION "version 1.4.0W"
+#define VERSION "Genesis Plus GX 1.4.0 (WII)"
 #else
-#define VERSION "version 1.4.0G"
+#define VERSION "Genesis Plus GX 1.4.0 (GCN)"
 #endif
 
 #define osd_input_Update() gx_input_UpdateEmu()
